server/src/web/broadcast.rs

```diff
@@ -0,0 +1,137 @@
+
+pub enum BroadcastEvent {
+    BuildStart,
+    BuildFinish,
+    Log(String)
+}
+
+pub struct Broadcast {
+    subscriptions: Mutex<HashMap<String, Vec<tokio::sync::mpsc::Sender<sse::Event>>>>,
+    // cache contains build logs for packages which are currently building
+    cache: Mutex<HashMap<String, Vec<String>>>,
+    db: Pool<Sqlite>
+}
+
+impl Broadcast {
+    pub fn new(db: Pool<Sqlite>) -> Arc<Self> {
+        let broadcast = Arc::new(Self { 
+            subscriptions: Mutex::new(HashMap::new()),
+            cache: Mutex::new(HashMap::new()),
+            db
+        });
+        Broadcast::spawn_ping(broadcast.clone());
+        broadcast
+    }
+
+    fn spawn_ping(this: Arc<Self>) {
+        tokio::task::spawn(async move {
+            let mut interval = tokio::time::interval(Duration::from_secs(10));
+
+            loop {
+                interval.tick().await;
+                this.remove_stale_connections().await;
+            }
+        });
+    }
+
+    /// remove all connections for which the ping broadcast fails to avoid unnecessary broadcasts
+    async fn remove_stale_connections(&self) {
+        let mut subscriptions = self.subscriptions.lock().await;
+
+        *subscriptions = join_all(
+            subscriptions.iter().map(|(package, receivers)| async {
+                let receivers = join_all(
+                    receivers.iter().map(|recv| async {
+                        let event = sse::Event::Data(sse::Data::new(Utc::now().timestamp_millis().to_string()).event(PING_EVENT));
+                        recv.send(event).await.ok().map(|_| recv.clone())
+                    })
+                ).await.into_iter().filter_map(|r| r).collect::<Vec<_>>();
+                if receivers.len() > 0 {
+                    Some((package.clone(), receivers))
+                } else {
+                    None
+                }
+            })
+        ).await.into_iter().filter_map(|e| e).collect::<HashMap<_, _>>();
+    }
+
+    /// subscribe to all package events
+    pub async fn subscribe(&self, package: String) -> actix_web::Result<Sse<InfallibleStream<ReceiverStream<sse::Event>>>> {
+        let pkg = package.to_lowercase();
+        let (tx, rx) = tokio::sync::mpsc::channel::<sse::Event>(10);
+        let mut subscriptions = self.subscriptions.lock().await;
+        let mut receivers = subscriptions.get(&pkg).cloned().unwrap_or_default();
+        debug!("added new receiver for package {pkg}");
+        receivers.push(tx.clone());
+        subscriptions.insert(pkg.clone(), receivers);
+
+        let cache = self.cache.lock().await;
+        // should there be logs in the cache then there is currently a build running and we want to return those logs
+        // otherwise we return the latest logs from the database
+        if let Some(logs) = cache.get(&pkg) {
+            let event = sse::Event::Data(sse::Data::new(logs.join("")).event(LOG_EVENT));
+            tx.send(event).await.ok();
+        } else {
```

Comment:
I think we should not read from the database here, but just wait for the next build. The subscribed logs should always be live. (See my comment in the cli)


This way we also don't need to pass the db into this struct, keeping things a tiny bit simpler.

Reply:
Actually, the log stream stops when the container exits. Therefore, we can indeed remove the channel.